### PR TITLE
Move core provider factory functionality to the provider package.

### PR DESCRIPTION
**Pull request by benashz**
Currently, it is not possible to test `meta.NewProviderMeta()` since it depends the `vault.Provider()`, which creates a cyclical dependency. It also makes sense that the provider factory would be in the lower level 'provider' package.

### CODE REVIEW

File: internal/provider/provider.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+		panic(err)
+	}
+
+	resourcesMap, err := parse(resourceRegistry)
+	if err != nil {
+		panic(err)
+	}
+
+	for _, m := range resourceMap {
+		MustAddSchemaResource(m, resourcesMap, nil)
```
Comment (vinay-gopalan) [inline internal/provider/provider.go:62]: should this be
```suggestion
		MustAddSchemaResource(m, resourceMap, nil)
```

Comment (benashz) [reply internal/provider/provider.go:62]: Variable names a bit confusing, I think what we have there is correct. It comes from:
https://github.com/hashicorp/terraform-provider-vault/pull/1884/files#diff-7e69de666c9006fa0a0fb8fe4e4f8c418acce7c555f69f8087d88d16b434f2c1L59

Comment (vinay-gopalan) [reply internal/provider/provider.go:62]: Ah you're right, thanks for clarifying

**Review by vinay-gopalan — APPROVED** [review]
LGTM!

File: internal/provider/provider.go (hunk 1 above)
Comment (benashz) [reply internal/provider/provider.go]: Make the variable names a little less confusing: f663882da8043ae07f30111e8b206349b1e2b592

Comment (vinay-gopalan) [reply internal/provider/provider.go]: Nice!

**Pull request merged**